### ISSUE TITLE
Bridging MQTT Connections  Handshake Failure {bad_cert,hostname_check_failed}

### ISSUE THREAD

**Issue by elap4** [vernemq/vernemq#1439]
### Environment

- VerneMQ Version: 
 Raspberry Pi 4:  1.10.0+build.7.refc432ed7
 Remote broker 1.10.1
- OS:  Raspbian Linux 10 Buster
- Erlang/OTP version (if building from source): 22
- VerneMQ configuration (vernemq.conf) or the changes from the default

```
plugins.vmq_bridge = on
vmq_bridge.ssl.sbr0 = 192.168.1.100:8883
vmq_bridge.ssl.sbr0.topic.1 = bridgedtopic
vmq_bridge.ssl.sbr0.cafile = /path/to/ca/certificate.pem
```
- Cluster size/standalone: 1 (standalone)

### Expected behavior

Client connects to remote broker using TLS

### Actual behaviour

Client broker does not connect to remote broker 

`connection to "192.168.1.100":8883 failed due to {tls_alert,{handshake_failure,"TLS client: In state certify at ssl_handshake.erl:1768 generated CLIENT ALERT: Fatal - Handshake Failure\n {bad_cert,hostname_check_failed}"}}`

### Additional Information
A non TLS bridge can be set up between vernemq nodes
Setting vmq_bridge.ssl.sbr0.insecure = on allows the connection to happen

- I have been able to bridge a mosquitto broker to vernemq using the CA certificate

- Clients can connect to the host broker using the CA certificate

- The remote broker/host broker has a certificate signed by the CA and the common name is set to the IP address used to connect 192.168.1.100

**Comment by ioolkos** [comment]
Is certificate.pem a chain of multiple certs?

**Comment by elap4** [comment]
Single CA used to sign all broker and client certs, self signed
I have tried regenerating also

**Comment by ioolkos** [comment]
There's now an option to configure certs depth for the bridge client: `vmq_bridge.ssl.sbr0.depth = 4` for cert chains.

**Comment by elap4** [comment]
Looks like I need to rebuild with the latest changes as its unsupported in my current installation

I think my current depth should be at 1 though as the CA has directly signed the broker/client certs 
Does this seem like a likely solution?

**Comment by ioolkos** [comment]
Yeah, you're probably right, might not help in this case. Please keep this issue updated if you find any more insights.

**Comment by elap4** [comment]
I did have a look through the vmq_bridge module but I'm not familiar with Erlang
I'm not sure how TLS is integrated 
Based on the error messages I've got through various attempts (giving it a corrupt certificate, wrong IP addresses)  I'm assuming its where the certificate is verified but I was unable to locate this code

If you know where this is/how it happens, it might shed some light on what steps to take next

**Comment by mrceyhun** [comment]
Hi all, I'm facing with similar issue while bridging an instance using its self-signed certificate with IP address.
vernemq version: 1.12.4
vernemq conf:
```
vmq_bridge.ssl.br0 = x.x.x.x:8883
vmq_bridge.ssl.br0.username = x
vmq_bridge.ssl.br0.password = x
vmq_bridge.ssl.br0.cleansession = on
vmq_bridge.ssl.br0.client_id = auto
vmq_bridge.ssl.br0.keepalive_interval = 60
vmq_bridge.ssl.br0.restart_timeout = 10
vmq_bridge.ssl.br0.max_outgoing_buffered_messages = 10000
vmq_bridge.ssl.br0.cafile = /vernemq/etc/certs/tls.crt
vmq_bridge.ssl.br0.depth = 10
vmq_bridge.ssl.br0.insecure = off
vmq_bridge.ssl.br0.try_private = off
```

Vernemq errors:
```
[error] connection to "x.x.x.x":8883 failed due to {tls_alert,{handshake_failure,"TLS client: In state certify at ssl_handshake.erl:1768 generated CLIENT ALERT: Fatal - Handshake Failure\n {bad_cert,hostname_check_failed}"}}
[error] connection failed due to server_not_found
```

When I try openssl s_client in the vernemq instance using the cert, output is like:
```
vernemq@19ba90d8278e:~$ openssl s_client -CAfile /vernemq/etc/certs/tls.crt  -connect x.x.x.x:8883
CONNECTED(00000003)
Can't use SSL_get_servername
depth=1 CN = x, O = x, L = x, ST = x, C = x
verify return:1
depth=0 CN = x, O = x, L = x, ST = x, C = x
verify return:1
---
Certificate chain
 0 s:CN = x, O = x, L = x, ST = x, C = x
   i:CN = x Root CA, O = x, L = x, ST = x, C = x
 1 s:CN = x Root CA, O = x, L = x, ST = x, C = x
   i:CN = x Root CA, O = x, L = x, ST = x, C = x
---
```

Maybe it can contribute solving the issue.

**Comment by ioolkos** [comment]
@mrceyhun Thanks... I think you could set `vmq_bridge.ssl.br0.insecure = on` except that, well, it's insecure.
Possibly also `vmq_bridge.ssl.bro0.customize_hostname_check = 'https'` which switches to HTTPS style hostname_checks allowing for wildcards.
Other than that, I'm not sure what could help with IPs and self-signed certs.

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by mrceyhun** [comment]
Thanks @ioolkos , I didn't know there was an option `customize_hostname_check` in MQTT bridging. Most probably because of my VerneMQ version, `insecure = on` setting caused vernemq config validation error. `customize_hostname_check` (no typo of `bro0`)  did not help either.

Anyway, I handled it using insecure 1883 port.

**Comment by ioolkos** [comment]
@mrceyhun allright, although even version `1.12.4` should have those options.
As a reminder, you can validate and debug the conf file by `vernemq config generate -l debug`

---
👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.
